lib/path/default.nix

```diff
@@ -92,8 +95,114 @@ let
     # An empty string is not a valid relative path, so we need to return a `.` when we have no components
     (if components == [] then "." else concatStringsSep "/" components);
 
+  # Takes a Nix path value and deconstructs it into:
+  # - root: The filesystem root of the path, generally `/`
+  # - components: All the path components of the path
+  #
+  # This is a bit like `splitString "/" (toString path)` but safer
+  # because it can distinguish different filesystem roots
+  deconstructPath =
+    let
+      go = components: path:
+        # If the parent of a path is the path itself, then it's a filesystem root
+        if path == dirOf path then { root = path; inherit components; }
+        else go ([ (baseNameOf path) ] ++ components) (dirOf path);
+    in go [];
+
 in /* No rec! Add dependencies on this file at the top. */ {
 
+  /*
+  Whether the second path is a descendant of the first path, or equal to it.
```

Comment:
Paths do represent trees though, I don't think anybody would be confused by that. In contrast, calling these functions `isPrefixOf` or `hasPrefix` has a good chance of confusing users with the string-based `lib.strings.hasPrefix`, `lib.strings.hasSuffix`, etc.

Reply:
> Paths do represent trees though

I'm genuinely confused now. How is a path not a sequence but a tree?

And while common parlance conflates the two, that kind of fuzzy thinking leads to fuzzy thinking that we really want to avoid.

Reply:
I'm not saying a path _is_ a tree, I'm saying it represents one. E.g. if you import /foo into the store is will also import /foo/bar because it's a descendant.

But I now see that this doesn't really hold when the paths don't exist, a parent-descendant relationship doesn't make sense then. So yes agreed this naming isn't ideal.

Reply:
Okay, you did not say it *is* one in terms of equivalence, but it also doesn't represent one. A path *leads* to a file system object, which *can* be a tree – or a single file. That is something very different than representation. A string of some shape may represent a path, or a drawing of some sort may represent a tree.

Reply:
We're getting down to definitions now, arguably the source of most arguments. Here's the definitions for "represent" in the Merriam-Webster dictionary: https://www.merriam-webster.com/dictionary/representing

> 2.  to serve as a sign or symbol of 
>     > the flag represents our country

This is the definition I was using.

But you're right, in programming we often use the term "represent" to mean different representations of the same thing, you can often convert from one representation to another and back without information loss. I'm not sure which dictionary definition fits that though.

Reply:
We could use `startsWith /foo /foo/bar` insteaed, which mirrors Rust's [`starts_with`](https://doc.rust-lang.org/std/path/struct.Path.html#method.starts_with).

Only then it's a bit unclear what the `isProperDescendantOf` version should be. Maybe we can just leave that function out, users can always have an additional `!=` check when they need that version.

Another problem then would be what [`lib.path.difference`](https://github.com/NixOS/nixpkgs/pull/209375) should return, `commonStart` sounds odd, maybe we could just use `prefix` then at the cost of some consistency.

---

I guess calling this `hasPrefix` might indeed be the best option locally, but I really don't like the potential confusion with `lib.strings.hasPrefix`, which notably has totally unexpected behavior for paths:

```
nix-repl> lib.strings.hasPrefix ./lib ./lib/cli.nix
false
```

And it's exported from `lib`:
```
nix-repl> lib.hasPrefix ./lib ./lib/cli.nix
false
```

We could perhaps change `hasPrefix` to give an error for path values, but that's a breaking change. Though I think that might be the best way forward because honestly most `lib.strings` functions don't make a lot of sense to call on paths directly, they all implicitly convert them to store paths.

Reply:
> But I now see that this doesn't really hold when the paths don't exist

Paths are always fictional things for which existence is irrelevant. It's the file system object it references that matters, and I have yet to come across a standard library that has a type of path that guarantees to be backed by a real file system object. Those objects are files or handles, and usually these objects forget the path. I don't see a problem with terminology that operates on a fictional tree of all paths.

I think we're all clever enough to find arguments for and against everything or weasel our way out. For example: If fictional isn't good, we can't have `hasPrefix` either, because it operates on a fictional set of possible prefixes that may have been used to construct the path.

I guess the real question is whether we can pick an abstraction that works reasonably well for all the operations we want to have on paths?

Reply:
@roberth I don't think anyone here is trying to weasel their way into being right, and instead we try to find the most convincing answer. 🙂

> I really don't like the potential confusion with `lib.strings.hasPrefix`

@infinisil We brought this up multiple times and while I see your concern, I consider it lower priority. The attribute-set-namespace explicitly says one thing is for strings, the other for paths. Relative paths are not regular strings. We should plaster that all over documentation. Yes, one can discard the namespace prefix on import, but you'd have to do it deliberately in each file if you follow best practices.

> We could perhaps change `[string.]hasPrefix` to give an error for path values, but that's a breaking change.

> most `lib.strings` functions don't make a lot of sense to call on paths directly, they all implicitly convert them to store paths.

Fully agree.

> Paths are always fictional things for which existence is irrelevant.

Exactly, although with existence we may mean two different things:
- existence of a path in the directory tree, in the graph-theoretic sense
- existence of a path in the sense of being written down in some representation

The former indeed only starts to matter once we want to traverse the path in the directory tree.
What we care about in this library is only the latter, and suddenly the conceptual model matters, and that is – sorry for being repetitive – a sequence of path components.

Reply:
> Yes, one can discard the namespace prefix on import, but you'd have to do it deliberately in each file if you follow best practices.

The thing is, people aren't following best practices! In the grand scheme of things, what we do here is highly irregular, no third party takes that much care writing their code.

In particular, the (arguably anti-)pattern of using `with lib;` is all over the place. If both `lib.path.hasPrefix` and `lib.hasPrefix` exist, better make sure to never forget the `path.` prefix, otherwise you'll get wrong results like the one I showed. And this is even more likely to happen because traditionally everything has been re-exported from `lib.*`, meaning people may assume that `lib.path.hasPrefix` also follows that, and sure enough `lib.hasPrefix` exists.

Because of that I don't think we can "just" have `lib.path.hasPrefix` be a thing. The only reasonably thing that I can see working is to deprecate support for passing paths to `lib.strings.hasPrefix`.

Additionally we could also deprecate `lib.hasPrefix` and require people to use `lib.strings.hasPrefix` instead, but this would break _a lot_ of code so I'd rather not unless/until we're sure that these `lib.*` are truly a bad idea.

Reply:
I now opened https://github.com/NixOS/nixpkgs/pull/221204 to implement my suggestion of deprecating `hasPrefix` for paths.

Reply:
I now switched this PR to use `hasPrefix` and `hasProperPrefix`, which I think is okay if the above PR is merged

Reply:
Weak typing was particularly bad for `hasPrefix`:

```
nix-repl> lib.hasPrefix ~/. ~/h/nixpkgs
^C^C^C^C^C^C^C^C^C^C^Z
[1]+  Stopped                 nix repl

$ kill -9 %%
```

Friends don't let friends add their home directory to the store ;)